jung-algorithms/src/main/java/edu/uci/ics/jung/algorithms/transformation/FoldingTransformer.java

```diff
@@ -52,274 +57,83 @@
      * contains edges <code>(a,c)</code> and <code>(c,b)</code> for at least
      * one vertex <code>c</code>.
      * 
      * <p>The vertices of the new graph are the same as the vertices of the
-     * appropriate partition in the old graph; the edges in the new graph are
-     * collections of the intermediate vertices <code>c</code>.
+     * appropriate partition in the input graph.  The edge values are the sets of
+     * nodes that connected the edge's endpoints in the input graph.
      * 
      * <p>This function will not create self-loops.
      * 
-     * @param <V> vertex type
+     * @param <N> vertex type
      * @param <E> input edge type
-     * @param g input k-partite graph
-     * @param p predicate specifying vertex partition
-     * @param graph_factory Supplier used to create the output graph 
-     * @return the result of folding g into unipartite graph whose vertices
-     * are those of the <code>p</code> partition of g
+     * @param g input graph
+     * @param nodes input node set
      */
-    public static <V,E> Graph<V, Collection<V>> foldKPartiteGraph(KPartiteGraph<V,E> g, Predicate<V> p, 
-            Supplier<Graph<V, Collection<V>>> graph_factory)
+	// TODO: consider providing ValueGraph/Network versions of this
+	// TODO: consider renaming this
+    public static <N> ValueGraph<N, Set<N>> foldToValueGraph(Graph<N> g, Set<N> nodes)
     {
-        Graph<V, Collection<V>> newGraph = graph_factory.get();
-
-        // get vertices for the specified partition, copy into new graph
-        Collection<V> vertices = g.getVertices(p);
-
-        for (V v : vertices)
-        {
-            newGraph.addVertex(v);
-            for (V s : g.getSuccessors(v))
-            {
-                for (V t : g.getSuccessors(s))
-                {
-                    if (!vertices.contains(t) || t.equals(v)) 
-                        continue;
-                    newGraph.addVertex(t);
-                    Collection<V> v_coll = newGraph.findEdge(v, t);
-                    if (v_coll == null)
-                    {
-                        v_coll = new ArrayList<V>();
-                        newGraph.addEdge(v_coll, v, t);
-                    }
-                    v_coll.add(s);
-                }
-            }
+    	Preconditions.checkArgument(g.nodes().containsAll(nodes),
+    			"Input graph must contain all specified nodes");
+    	ValueGraphBuilder<Object, Object> builder = g.isDirected()
+    			? ValueGraphBuilder.directed()
+    			: ValueGraphBuilder.undirected();
+    	MutableValueGraph<N, Set<N>> newGraph = builder
+    			.expectedNodeCount(nodes.size())
+    			.nodeOrder(g.nodeOrder())
+    			.build();
+
+        for (N node : nodes) {
+        	for (N s : g.successors(node)) {
+        		for (N t : g.successors(s)) {
+        			if (!nodes.contains(t) || t.equals(node)) {
+        				continue;
+        			}
+        			Set<N> intermediateNodes = newGraph.edgeValueOrDefault(node, t, new HashSet<N>());
```

Comment:
@jrtom I'm just wondering if a `LinkedHashSet` would be a better fit here than a `HashSet`, so that the element ordering of the set is guaranteed to be consistent. WDYT?